src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1031,8 +1031,15 @@ public SolutionState WithProjectAnalyzerReferences(ProjectId projectId, IEnumera
             // we changed, rather than creating an entire new generator driver from scratch and rerunning all generators, is cheaper
             // in the end. This was written without data backing up that assumption, so if a profile indicates to the contrary,
             // this could be changed.
-            var addedReferences = newProject.AnalyzerReferences.Except(oldProject.AnalyzerReferences).ToImmutableArray();
-            var removedReferences = oldProject.AnalyzerReferences.Except(newProject.AnalyzerReferences).ToImmutableArray();
+            //
+            // When we're comparing AnalyzerReferences, we'll compare with reference equality; AnalyzerReferences like AnalyzerFileReference
+            // may implement their own equality, but that can result in things getting out of sync: two references that are value equal can still
+            // have their own generator instances; it's important that as we're adding and removing references that are value equal that we
+            // still update with the correct generator instances that are coming from the new reference that is actually held in the project state from above.
+            // An alternative approach would be to call oldProject.WithAnalyzerReferences keeping all the references in there that are value equal the same,
+            // but this avoids any surprises where other components calling WithAnalyzerReferences might not expect that.
```

Comment:
Now that I'm aware of this gotcha, I'm questioning the generally goodness of having AnalyzerFileReferences be equal comparable in the first place, but since that'd be an API breaking change I'm doing this for 17.4 and 17.5 Preview 1; we can take as a follow up item questioning the wisdom of all of this.

Reply:
oh yes.  we're on the same page.  i thin kwe should break this.